cmd/manager/main.go

```diff
@@ -68,125 +72,153 @@ const (
 )
 
 var (
-	// Cmd is the cobra command to start the manager.
-	Cmd = &cobra.Command{
+	configFile string
+	log        logr.Logger
+)
+
+func Command() *cobra.Command {
+	cmd := &cobra.Command{
 		Use:   "manager",
-		Short: "Start the operator manager",
-		Long: `manager starts the manager for this operator,
- which will in turn create the necessary controller.`,
-		Run: func(cmd *cobra.Command, args []string) {
-			execute()
+		Short: "Start the ECK operator",
+		PreRunE: func(cmd *cobra.Command, _ []string) error {
+			if configFile != "" {
+				viper.SetConfigFile(configFile)
+				if err := viper.ReadInConfig(); err != nil {
+					return fmt.Errorf("failed to read config file %s: %w", configFile, err)
+				}
+
+				viper.WatchConfig()
+				viper.OnConfigChange(func(_ fsnotify.Event) {
+					if log != nil {
+						log.Info("Detected an update to the configuration file. Restart the operator for the changes to take effect.")
```

Comment:
I think as a user I would expect ECK to automatically restart. This is the current behaviour for all the stack components so far (and more generally what happen in K8S with the reconcile loop pattern).

An other argument is that I'm not sure how restarting manually fits with a "gitops" approach.

Reply:
With GitOps, you'd create a dependency between the config map and the pod (trivially, this could be an annotation in the pod that gets updated with the hash of the config file) and that would cause the pod to be restarted automatically.

Unfortunately, the way the operator is structured, I don't see a way to "hot reload" the configuration. So then the question is how to force a restart. One way is to call `os.Exit` but that seems too crude and assumes that there's a restart mechanism available. The other one I can think of is having a liveness probe that starts failing when the config changes.  

Do you have any thoughts on achieving this nicely without complicating the operator too much?

Reply:
> Do you have any thoughts on achieving this nicely without complicating the operator too much?

The manager can be stopped through a [channel](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/internal.go#L438), I would try to explore this solution.

Reply:
I had a [go at it](https://github.com/elastic/cloud-on-k8s/pull/3570/commits/e3bed8b1acfb26ddd0ae80e14c4274fb723a6b11) and it kind of works. The issues with it are:
- Log level cannot be changed by that method (See #3593)
- Unmanaged goroutines in the code could start leaking because they don't have a proper shutdown mechanism

Reply:
I think we can just let the operator stop completely and delegate the restart of the container to the statefulset controller.
I think this is what would happen if the user stopped it itself.